internal/notifications/notifications.go

```diff
@@ -4,134 +4,495 @@ import (
+// Check if all WantRequests no longer want the block (ie it can be cancelled)
+func (wrm *WantRequestManager) shouldSendCancel(c cid.Cid) bool {
+	kwrs, ok := wrm.wrs[c]
+	if !ok {
+		return true
+	}
+
+	readyCount := 0
+	for wr := range kwrs {
+		if wr.cancelReady(c) {
+			readyCount++
+		} else if !wr.cancelled(c) {
+			// If there is a WantRequest that's not in the "ready to cancel"
+			// or "already cancelled" states, don't cancel yet
+			return false
+		}
+	}
+
+	// All WantRequests are either ready to cancel, or have already cancelled
+	// the want.
+	// Only send a cancel for the want if all WantRequests haven't already
+	// cancelled it.
+	return readyCount > 0
+}
+
+func (wr *WantRequest) Notifications() <-chan *Notification {
+	return wr.ntfns
+}
+
+// Called when a incoming message arrives that has blocks / HAVEs / DONT_HAVEs
+// for the keys in this WantRequest
+func (wr *WantRequest) receive(msg *IncomingMessage) *cid.Set {
+	wr.lk.Lock()
+	defer wr.lk.Unlock()
+
+	wanted := cid.NewSet()
+	if wr.closed {
+		return wanted
+	}
+
+	// Filter incoming message for blocks / HAVEs / DONT_HAVEs that this
+	// particular WantRequest is interested in
+	fmsg := &IncomingMessage{
+		From:      msg.From,
+		Blks:      wr.filterBlocks(msg.Blks),
+		Haves:     wr.filterKeys(msg.Haves),
+		DontHaves: wr.filterKeys(msg.DontHaves),
+	}
+
+	// Work out which blocks are still wanted by this WantRequest (ie this is
+	// the first time the block was received)
+	for _, b := range fmsg.Blks {
+		c := b.Cid()
+		if state, ok := wr.ks[c]; ok && state == ReqKeyWanted {
+			wr.ks[c] = ReqKeyUnwanted
+			wanted.Add(c)
+		}
+	}
+
+	// TODO: Actually we do want to keep sending notifications even after all
+	// blocks are received, so that the session can hear about other peers that
+	// have the blocks it's interested in
+
+	// If all blocks have been received already, no need to send the session
+	// any more notifications
+	if wr.allReceived {
+		return wanted
+	}
+
+	// Check if there are any blocks that the WantRequest is still waiting
+	// to receive
+	remaining := false
+	for _, state := range wr.ks {
+		if state == ReqKeyWanted {
+			remaining = true
+			break
+		}
+	}
+
+	// Send notification with the message and the list of wanted blocks
+	// (as opposed to duplicates)
+	wr.ntfns <- &Notification{fmsg, wanted}
+
+	// All blocks have been received, so close the notifications channel
+	if !remaining {
+		close(wr.ntfns)
+		wr.allReceived = true
+	}
+
+	return wanted
+}
+
+// Called when the Session has finished processing a received block
+func (wr *WantRequest) OnReceived(ks []cid.Cid) {
+	wr.lk.Lock()
+
+	if wr.closed {
+		wr.lk.Unlock()
+		return
+	}
+
+	for _, c := range ks {
+		if state, ok := wr.ks[c]; ok && state != ReqKeyCancelled {
+			wr.ks[c] = ReqKeyCancelReady
+		}
+	}
+
+	wr.lk.Unlock()
+
+	wr.wrm.onCancelReady(ks)
+}
+
+// Called when the request or session has completed
+func (wr *WantRequest) OnClosing() []cid.Cid {
```

Comment:
Can we just call this `Remaining()`? `On*` names should only be used as a last resort as they say absolutely nothing about what the function actually _does_.